src/VisualStudio/Core/Def/Telemetry/VisualStudioTelemetryLogManager.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.Telemetry;
+using Microsoft.VisualStudio.Telemetry;
+
+namespace Microsoft.VisualStudio.LanguageServices.Telemetry
+{
+    internal sealed class VisualStudioTelemetryLogManager
+    {
+        private readonly Dictionary<FunctionId, VisualStudioTelemetryLog> _logs;
+        private readonly TelemetrySession _session;
+        private readonly ILogger _telemetryLogger;
+        private readonly object _lock;
```

Comment:
comemnts about lock.  and move lock/lockees to same section of code.

Reply:
What do you think about something like the following instead of using a lock directly? I remember Roslyn had some fancy interlocked update methods, but I'm not finding anything that would help directly.

        private ImmutableDictionary<FunctionId, VisualStudioTelemetryLog> _immutableLogs = ImmutableDictionary<FunctionId, VisualStudioTelemetryLog>.Empty;

        public ITelemetryLog? GetLog(FunctionId functionId)
        {
            if (!_session.IsOptedIn)
                return null;

            return ImmutableInterlocked.GetOrAdd(ref _immutableLogs, functionId, new VisualStudioTelemetryLog(_telemetryLogger, functionId));
        }

Reply:
Went ahead and changed over to immutable members. Probably slightly more allocations, but it's cleaner and easier to reason about (and fewer chances of lock contention).